src/lib/server/textGeneration/tools.ts

```diff
@@ -0,0 +1,187 @@
+import { ToolResultStatus, type ToolCall, type ToolResult } from "$lib/types/Tool";
+import { v4 as uuidV4 } from "uuid";
+import JSON5 from "json5";
+import type { BackendTool, BackendToolContext } from "../tools";
+import {
+	MessageToolUpdateType,
+	MessageUpdateStatus,
+	MessageUpdateType,
+	type MessageUpdate,
+} from "$lib/types/MessageUpdate";
+import type { TextGenerationContext } from "./types";
+
+import { allTools } from "../tools";
+import directlyAnswer from "../tools/directlyAnswer";
+import websearch from "../tools/web/search";
+import { z } from "zod";
+import { logger } from "../logger";
+import { toolHasName } from "../tools/utils";
+import type { MessageFile } from "$lib/types/Message";
+import { mergeAsyncGenerators } from "$lib/utils/mergeAsyncGenerators";
+
+function makeFilesPrompt(files: MessageFile[]): string {
+	if (files.length === 0) return "";
+
+	const stringifiedFiles = files
+		.map((file, idx) => `  - fileIndex ${idx} | ${file.name} (${file.mime})`)
+		.join("\n");
+	return `The user attached ${files.length} file${
+		files.length === 1 ? "" : "s"
+	}:\n${stringifiedFiles}`;
+}
+
+export function pickTools(
+	toolsPreference: Record<string, boolean>,
+	isAssistant: boolean
+): BackendTool[] {
+	// if it's an assistant, only support websearch for now
+	if (isAssistant) return [directlyAnswer, websearch];
+
+	// filter based on tool preferences, add the tools that are on by default
+	return allTools.filter((el) => {
+		if (el.isLocked && el.isOnByDefault) return true;
+		return toolsPreference?.[el.name] ?? el.isOnByDefault;
+	});
```

Comment:
```suggestion
	return allTools.filter((el) => (el.isLocked && el.isOnByDefault) ?? toolsPreference?.[el.name] ?? el.isOnByDefault);
```

Reply:
I think that's less readable than the existing

Reply:
closing it then